### ISSUE TITLE
Remove StaticAssert from Contributor Guide/Best Practices

### ISSUE THREAD

**Issue by Lastique** [boostorg/website-v2-docs#188]
https://www.preview.boost.org/doc/contributor-guide/best-practices.html#_beneficial_dependencies

Boost.StaticAssert should probably be removed from the list. Majority of Boost is using C++11 or later as the baseline now, so core language `static_assert` should be preferred.

**Comment by vinniefalco** [comment]
I'm not sure I agree. `BOOST_STATIC_ASSERT` provides the 1-argument overload which allows omission of the string, a convenience I rather enjoy.

**Comment by Lastique** [comment]
Although I admit I sometimes omit the message in `static_assert` (in C++17) myself, this is often a result of laziness or time constraints rather than a deliberate decision that improves the resulting code quality. I think, most, if not all, `static_asserts` should have a message explaining the user what the problem is, should the assert fail. Thus I don't think including Boost.StaticAssert in the "Best Practices" page on the basis that it allows to omit the message is appropriate - because omitting the message is not the "best practice".

**Comment by vinniefalco** [comment]
That's your opinion but I don't see the evidence that static assertions which omit the message are not "best practices." For example no message is needed here:

```
static_assert( std::is_default_constructible_v<T> );
```

However, I can accept the argument that "while useful, static assertions without a message do not provide sufficient value to justify them as a beneficial dependency, especially since C++17 and later have no need for the macro."

One of Boost's strengths is that it offers newer language features for older compilers. BOOST_STATIC_ASSERT is the perfect example. We might consider if we should add a section "C++11 Users" (and "C++14 Users") which covers the Boost libraries that backfill new library features for older language versions. It would be in these sections where we might offer Boost.StaticAssert up.

**Comment by Lastique** [comment]
> That's your opinion but I don't see the evidence that static assertions which omit the message are not "best practices."

Of course, this is my opinion. This issue to begin with, is based on my opinion. You can disagree, and you can close it.

I think, `static_asserts` without messages are less user-friendly than those with. Simply because without the message the user has a harder time deciphering what failed and why. There are times when the root of the issue is plainly obvious, but even then, including a message if only to explain the problem in plain English is a friendly gesture towards the user.

> For example no message is needed here:
> 
> ```
> static_assert( std::is_default_constructible_v<T> );
> ```

Depending on where it's used, I'd put a message saying "Boost.Library: T must be default-constructible for this and that". This message conveys "who", "what" and "why" in plain English, which is beyond a random trait check.

**Comment by vinniefalco** [comment]
I'm not going to close it, if you want to remove it from best practices, I will not object to it.
